### PR TITLE
Stub out internal modules for tree DDS

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

This stubs out a lot of the structure for the tree DDS as documented in modules.md.

The architecture is indented to be usable long term, but the details of the actual interfaces are very much work in progress targeted ag getting a milestone 1 level demo working, and will likely be evolved significantly long term.

## Reviewer Guidance

The primary goal here is to get more surface area for the team to start coding on (maximize velocity), and keep it organized to minimize tech debt.

It is not indented that the included code is robust (or even working), but rather that it serves as a good starting point for people to start working on.

To maximize development parallelization, interfaces are being kept simple (performance work in the future may have to complicate a few of them), and everything is designed to be easy to unit test / develop in parallel.

If you can spot areas that will be difficult to unit test or modify/replace in the future, that would be useful.

### CODE REVIEW

File: packages/dds/tree/src/object-forest/objectForest.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        this.root = root;
+        this.state = ITreeSubscriptionCursorState.Current;
+        this.indexStack.push(index);
+        this.siblings = this.forest.getRoot(root);
+    }
+
+    getNode(): ObjectNode {
+        assert(this.state === ITreeSubscriptionCursorState.Current, "Cursor must be current to be used");
+        assert(this.parentStack.length > 0, "Cursor must be current to be used");
+        return this.parentStack[this.parentStack.length - 1];
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/object-forest/objectForest.ts:223]: Do we not expect to be able to call `getNode` after `set` is called? Why not?

File: packages/dds/tree/src/rebase/rebaser.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+// TODO: managing the types with these is not working well (inferring any for methods in Rebaser). Do something else.
+
+export type ChangeFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
+    TChangeRebaser extends ChangeRebaser<infer TChange, any, any> ? TChange : never;
+
+export type FinalFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
+    TChangeRebaser extends ChangeRebaser<any, infer TFinal, any> ? TFinal : never;
+
+export type SetFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/rebase/rebaser.ts:75]:
```suggestion
export type StateFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any, any, any>> =
```
You'll need to update usages as well.

File: packages/dds/tree/src/forest/snapshot.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    tryGet(id: NodeId): ITreeCursor | undefined;
+
+    /**
+     * @returns the parent of `id`. Should not be used if there is no node with id or if id refers to the root node.
+     */
+    getParent(id: NodeId): ChildLocation;
+
+    /**
+     * @returns undefined iff root, otherwise the parent of `id`.
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/forest/snapshot.ts:43]: you must do the access using a cursor from the current forest, so thats not possible. I've clarified/fixed this a bit.

File: packages/dds/tree/src/object-forest/objectForest.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/object-forest/objectForest.ts]: I expect that to work,

File: packages/dds/tree/src/rebase/rebaser.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/rebase/rebaser.ts:75]: It was the name of the identifier in the infer block that was wrong here.

File: packages/dds/tree/src/feature-libraries/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Feature Libraries
+
+Concrete implementations of abstractions use to parameterize `SharedTreeCore`, or any other entry points into `tree`.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/README.md:3]:
```suggestion
Concrete implementations of abstractions used to parameterize `SharedTreeCore`, or any other entry points into `tree`.
```

File: packages/dds/tree/src/feature-libraries/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+`SharedTreeCore` should be usable with alternant versions of anything withing this library.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/README.md:4]:
```suggestion
`SharedTreeCore` should be usable with alternative versions of anything withing this library.
```

File: packages/dds/tree/src/feature-libraries/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+This directory will end up containing a wide variety of implementations, including:
+
+- Definitions of fields kinds for schema
+- Schema languages (which support some of these field kinds)
+- Families of changes/edits which can be applies for various field kinds
+- ChangeRebaser's for these change families
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/README.md:11]:
```suggestion
- `ChangeRebaser` implementations for these change families
```

**Review by yann-achard-MS — APPROVED** [review]

**Pull request merged**